### DISCUSSION TITLE
Using an asset handle inside an async task?

### DISCUSSION THREAD

**Discussion by viridia** [bevyengine/bevy#9277] [Q&A]
I have an async task which is used for terrain generation. The task generates new terrain patches as the player moves around the world, allowing for worlds of unbounded size.

The terrain generator depends on an asset that contains data used during the generation process. This asset only needs to be loaded once at startup, and remains resident in memory for the life of the program.

I'm having trouble figuring out how to code this. If I store the asset handle in a resource, the borrow checker complains when I try to use the handle inside the closure for the task.

Also, I would like the task closure to be able to wait until the asset is loaded, but there seems to be no way to `await` for an asset. An alternative is to delay spawning the task until the asset is loaded, but that makes the logic more complex and doesn't generalize to other use cases very well. (In the future I might need more assets inside of tasks, and having to delay spawning the task until all those assets are loaded makes the code very complicated. It's simpler to just await for each loadable resource.)

Can anyone provide me with a code example showing how to do this?

**Comment by RCoder01** [comment]
I may be misunderstanding your question, but handles and the asset server are inexpensively cloneable:
```
#[derive(Debug, Resource)]
struct MyHandle(Handle<Foo>);

fn terrain_gen(my_handle: Res<MyHandle>, server: Res<AssetServer>) {
    let handle = my_handle.0.clone();
    let server = server.as_ref().clone();
    ComputeTaskPool::get().spawn(async move {
        pre_calculation().await; // do some calculation before
        // from now on, asset is needed
        loop { // idk if this blocks the compute thread
            match server.get_load_state(&handle) {
                LoadState::Loaded => { break; },
                _ => {}, // unloaded, loading, or failed
            }
        }
        async_foo(handle).await; // use handle somehow
    }).detach();
}
```

**Reply by viridia** [reply]
That solves part of the problem, thanks. Although I don't like the idea of busy-waiting the thread until the asset is loaded...Maybe I'll try calling `thread::yield()` or something.

**Reply by viridia** [reply]
So, your example works - up to a point. However, when I actually try to use the handle:

```
pub fn compute_ground_meshes(
    mut commands: Commands,
    mut query: Query<(Entity, &mut Parcel)>,
    asset_server: Res<AssetServer>,
    terrain_shapes: Res<TerrainShapesHandle>,
    shape_table: Res<Assets<TerrainShapesResource>>,
) {
    let pool = AsyncComputeTaskPool::get();

    for (entity, mut parcel) in query.iter_mut() {
        match parcel.status {
            // Spawn task to compute mesh
            ParcelStatus::New | ParcelStatus::Waiting => {
                let handle = terrain_shapes.handle.clone();
                let server = asset_server.as_ref().clone();
                let shape_table_resource = shape_table.as_ref().clone();
                parcel.status = ParcelStatus::Building;
                let task = pool.spawn(async move {
                    loop {
                        match server.get_load_state(&handle) {
                            LoadState::Loaded => {
                                break;
                            }

                            LoadState::Failed => {
                                panic!("Asset load failed");
                            }

                            _ => {
                                std::thread::yield_now();
                            }
                        }
                    }

                    let sh = shape_table.get(&handle);
```

I get:

```
borrowed data escapes outside of function
`shape_table` escapes the function body here
```

**Reply by hymm** [reply]
You'll have to clone your data out of `Res<Assets<_>>`. The spawned task can't hold any references to the world. I'd probably move checking the handle load state into the system and spawn the task once the data is loaded. If cloning is too expensive, you might try putting an `Arc<Mutex<_>>` into your asset.

**Reply by RCoder01** [reply]
The borrow error is because the move closure is trying to take hold of `Res<Assets<TerrainShapesResource>>` but I think the `compute_ground_meshes` is only given a `Res` bounded by the duration of the system running, whereas the async compute thread can and will run for significantly longer.

It's a pretty gross solution, but maybe have an `Arc<Mutex<Vec<TerrainShapesResource>>>` (or `RwLock`) shared between the world systems and the async compute thread, and another system that checks for when all the assets are loaded and puts them into the `Arc<Mutex<Vec<TerrainShapesResource>>>`.

```
fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, start_asset_loading)
        .add_systems(Update, check_asset_progress)
        .run();
}

#[derive(Resource)]
struct TerrainShapesHandle(Handle<TerrainShapesResource>);

#[derive(Resource, Clone, Default)]
struct SharedAssets(Arc<Mutex<Vec<TerrainShapesResource>>>);

fn start_asset_loading(mut commands: Commands, server: Res<AssetServer>) {
    commands.insert_resource(TerrainShapesHandle(server.load("foo")));
    let shared = SharedAssets::default();
    commands.insert_resource(shared.clone());
    AsyncComputeTaskPool::get().spawn(async move {compute(shared).await}).detach();
}

fn check_asset_progress(server: Res<AssetServer>, finished: Local<bool>, handle: Res<TerrainShapesHandle>, assets: ResMut<SharedAssets>, shapes: ResMut<Assets<TerrainShapesResource>>) {
    if !finished.deref() {
        if let LoadState::Loaded = server.get_load_state(handle.deref()) {
            assets.0.lock().push(shapes.remove(handle.deref().clone()).unwrap());
            *finished = true;
        }
    }
}

async fn compute(asset: SharedAssets) {
    loop {
        match asset.0.try_lock() {
            Ok(lock) => {
                if !lock.is_empty() {
                    break;
                }
            }
        }
        // not sure if there's a way to signal to the async executor to pause execution here
    }
}
```

**Reply by SkiFire13** [reply]
The right way to telling the executor to pause execution is to just yield to it, i.e. `.await` on something that doesn't immediately resolve to a value. There exist a bunch of `async` `yield_now` (`std::thread::yield_now` is not `async` and thus doesn't count) that do this, all very simple. I would suggest using [`futures_lite::future::yield_now`](https://docs.rs/futures-lite/1.13.0/futures_lite/future/fn.yield_now.html) since `futures_lite` is already a dependency of `bevy`.

**Reply by viridia** [reply]
@RCoder01 In your example, you have the `compute` function taking a parameter `asset: SharedAssets`. I wasn't aware that async tasks auto-inject parameters like systems do - is that documented anywhere?

**Reply by RCoder01** [reply]
`compute` isn't a system, it's just a normal function. Also, `SharedAsset` is a resource, but resources are just structs, which can be cloned and passed as normal. `compute` is called within ’start_asset_loading’ with a clone of the `SharedAsset` resource, which has an `Arc<Mutex<_>>` of the actual asset, so it can be accessed safely from multiple threads.

**Reply by viridia** [reply]
Also, I can't seem to figure out how to store a loaded asset into a mutex. I've tried doing it both ways: creating the mutex inside the asset loader, and outside at the point where we call assets.get(). Part of the problem is that the size of the asset is 56K and I'm trying to avoid cloning it.

The code just keeps getting more and more complicated :(

**Reply by viridia** [reply]
@RCoder01 Ah, thanks for the explanation, I misread the code :/

**Reply by RCoder01** [reply]
`assets.get()` only gives you a reference to the asset, to get an owned asset without cloning, you have to remove it from asset storage with `assets.remove()`

**Reply by viridia** [reply]
OK, that's progress. Here's what I am getting now:

```
if let Some(asset) = assets.remove(&handle.handle) {
  commands.insert_resource(TerrainShapesResource(Arc::new(Mutex::new(asset))));
}
```

```
cannot borrow data in dereference of `bevy::prelude::Res<'_, bevy::prelude::Assets<TerrainShapesAsset>>` as mutable
trait `DerefMut` is required to modify through a dereference, but it is not implemented for `bevy::prelude::Res<'_, bevy::prelude::Assets<TerrainShapesAsset>>`
```

**Reply by viridia** [reply]
This seems like it would be so much simpler if I could call assets.get() inside the task closure. I understand that the closure can't have references to the world - but is the asset server part of the world? I would think that an asset server would have a very long lifetime. The closure will have a strong handle, which will prevent the asset from being unloaded while the closure is running.

Also, what happens when TerrainShapes changes (due to terrain sculpting in the editor)? At this point, I'd want to re-run the task with the updated asset. But if the asset is stored in a resource, it means I have to write additional code to keep the resource in sync, so that the task can pick up the changes.

My concern here is that "TerrainShapes" isn't the only resource that I have which follows this pattern. My game (which I'm porting from three.js) has many different kinds of assets that are handled similarly - for example, there's a table of Biome types which is also a global asset and is also used in terrain mesh generation. Having to create two resources and two systems for every asset type seems like a lot of boilerplate - at least, when you compare it to the JavaScript version.

Apologies for all the questions, I'm fairly new to both Bevy and Rust.

**Reply by RCoder01** [reply]
I'm also pretty new to bevy, and I'm certain there's a better way to do all this (I have to do something similar in my project), but I don't know what it is.

The reason I don't think giving the task closure any sort of reference to the assets is possible is that `TaskPool::spawn` requires a "future: impl Future<Output = T> + Send + _'static_", and since the asset resource itself will live only as long as the world, I don't think any kind of `'static` reference to it is (safely) possible.

Also the reason this is much easier in JS is because JS is garbage collected and single threaded so there's no worry about leaking memory and less concurrent access memory concerns (maybe none at all, but I'm not a JS expert).

**Reply by viridia** [reply]
OK so I think I figured it out. The problem is that "Arc" can't derive Deserialize. This has to do with the fact that in the code example for custom resources, the type of the asset, and the raw type being loaded from a file are the same type. So that type derives all of the traits needed for both deserialization and for being an asset.

The problem comes when you need to do some kind of data conversion in the asset loader - in this case, wrapping an Arc and Mutex around the returned value. So you need two types (which I realized early on), but you also need to insure that those two types aren't trying to do too much by deriving from traits you don't need. So the "source" type needs the Deserialize trait, while the "asset" type needs the TypeUuid and TypePath traits. If you accidentally cut & paste such that your asset type is deriving deserialize, it won't work because Arc can't support that trait - but it doesn't need it either.

The other thing I realized is "don't poll the asset loading status, use asset events". That is, if you have a system which is checking whether the asset has finished loading, use an event reader, because this will not only tell you when the asset is loaded, it will also tell you when it has changed - which allows you to update the consumers of that asset and mark them for rebuild.
